File: services/data-warehouse/app/connectors/connector_factory.py
from enum import Enum
from typing import TypeVar, Union, Optional, Generic
from app.connectors.s3connector import S3Connector, S3ConnectorConfig

T = TypeVar('T')

class ConnectorType(Enum):
    S3 = "S3"

# Add unions for future connector types
Connector = Union[S3Connector[T]]

class ConnectorFactory(Generic[T]):
    @staticmethod
    def create(
        connector_type: ConnectorType,
        config: Optional[Union[S3ConnectorConfig]] = None
    ) -> Connector[T]:
        if connector_type == ConnectorType.S3:
            return S3Connector[T](config or S3ConnectorConfig())
        else:
            raise ValueError(f"Unknown connector type: {connector_type}")

File: services/sync-base/app/utilities/connectorFactory.ts
import { S3Connector } from "./s3connector";

export enum ConnectorType {
    S3 = "S3",
}

// Add unions for future connector types
export type Connector = S3Connector;

export function createConnector(type: ConnectorType): Connector {
    switch (type) {
        case ConnectorType.S3:
            return new S3Connector();
        default:
            throw new Error(`Unknown connector type: ${type}`);
    }
}